src/thread.ts

```diff
@@ -1,142 +1,873 @@
+
+        const { channelData } = this.state.getLatestValue();
+
+        if (!channelData || event.channel.cid !== channelData.cid) return;
+
+        this.state.patchedNext('isStateStale', true);
+      }).unsubscribe,
+    );
+
+    this.unsubscribeFunctions.add(
+      this.client.on('message.new', (event) => {
+        const currentUserId = this.client.user?.id;
+        if (!event.message || !currentUserId) return;
+        if (event.message.parent_id !== this.id) return;
+
+        const { isStateStale } = this.state.getLatestValue();
+
+        if (isStateStale) return;
+
+        if (this.failedRepliesMap.has(event.message.id)) {
+          this.failedRepliesMap.delete(event.message.id);
+        }
+
+        this.upsertReplyLocally({
+          message: event.message,
+          // deal with timestampChanged only related to local user (optimistic updates)
+          timestampChanged: event.message.user?.id === currentUserId,
+        });
+
+        if (event.message.user?.id !== currentUserId) this.incrementOwnUnreadCount();
+        // TODO: figure out if event.user is better when it comes to event messages?
+        // if (event.user && event.user.id !== currentUserId) this.incrementOwnUnreadCount();
+      }).unsubscribe,
+    );
+
+    this.unsubscribeFunctions.add(
+      this.client.on('message.read', (event) => {
+        if (!event.user || !event.created_at || !event.thread) return;
+        if (event.thread.parent_message_id !== this.id) return;
+
+        const userId = event.user.id;
+        const createdAt = event.created_at;
+        const user = event.user;
+
+        // FIXME: not sure if this is correct at all
+        this.state.next((current) => ({
+          ...current,
+          read: {
+            ...current.read,
+            [userId]: {
+              last_read: createdAt,
+              lastReadAt: new Date(createdAt),
+              user,
+              // TODO: rename all of these since it's formatted (find out where it's being used in the SDK)
+              unread_messages: 0,
+              // TODO: fix this (lastestReplies.at(-1) might include message that is still being sent, which is wrong)
+              last_read_message_id: 'unknown',
+            },
+          },
+        }));
+      }).unsubscribe,
+    );
+
+    const handleMessageUpdate = (event: Event<Scg>) => {
+      if (!event.message) return;
+
+      if (event.hard_delete && event.type === 'message.deleted' && event.message.parent_id === this.id) {
+        return this.deleteReplyLocally({ message: event.message });
+      }
+
+      this.updateParentMessageOrReplyLocally(event.message);
+    };
+
+    ['message.updated', 'message.deleted', 'reaction.new', 'reaction.deleted'].forEach((eventType) => {
+      this.unsubscribeFunctions.add(this.client.on(eventType, handleMessageUpdate).unsubscribe);
+    });
+  };
+
+  public deregisterSubscriptions = () => {
+    this.unsubscribeFunctions.forEach((cleanupFunction) => cleanupFunction());
```

Comment:
Shall we `this.unsubscribeFunctions.clear()` here?

Reply:
Yes, good point. :)